L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,390 @@
+  // be used as the name.
+  //
+  // @param identity_cert_name the name of identity key-cert pairs being set.
+  void set_identity_cert_name(const std::string& identity_cert_name);
+  // Sets the Tls session key logging configuration. If not set, tls
+  // session key logging is disabled. Note that this should be used only for
+  // debugging purposes. It should never be used in a production environment
+  // due to security concerns.
+  //
+  // @param tls_session_key_log_file_path: Path where tls session keys would
+  // be logged.
+  void set_tls_session_key_log_file_path(
+      const std::string& tls_session_key_log_file_path);
+  // Sets the certificate verifier used to perform post-handshake peer identity
+  // checks.
+  void set_certificate_verifier(
+      std::shared_ptr<CertificateVerifier> certificate_verifier);
+
+  // Sets the crl provider, see CrlProvider for more details.
+  void set_crl_provider(std::shared_ptr<CrlProvider> crl_provider);
+
+  // Sets the minimum TLS version that will be negotiated during the TLS
+  // handshake. If not set, the underlying SSL library will use TLS v1.2.
+  // @param tls_version: The minimum TLS version.
+  void set_min_tls_version(grpc_tls_version tls_version);
+  // Sets the maximum TLS version that will be negotiated during the TLS
+  // handshake. If not set, the underlying SSL library will use TLS v1.3.
+  // @param tls_version: The maximum TLS version.
+  void set_max_tls_version(grpc_tls_version tls_version);
+
+  // Sets a custom chain builder implementation that replaces the default chain
+  // building from the underlying SSL library.
+  void set_custom_chain_builder(std::shared_ptr<CustomChainBuilder>
+  chain_builder);
+}
+
+// Contains configurable options on the server side.
+class TlsServerCredentialsOptions final : public TlsCredentialsOptions {
+ public:
+  // Server side is required to use a provider, because server always needs to
+  // use identity certs.
+  explicit TlsServerCredentialsOptions(
+      std::shared_ptr<CertificateProviderInterface> certificate_provider)
+      : TlsCredentialsOptions() {
+    set_certificate_provider(certificate_provider);
+  }
+
+
+  // Sets option to request the certificates from the client.
+  // The default is GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE.
+  void set_cert_request_type(
+      grpc_ssl_client_certificate_request_type cert_request_type);
+}
+
+// Contains configurable options on the client side.
+// Client side doesn't need to always use certificate provider. When the
+// certificate provider is not set, we will use the root certificates stored
+// in the system default locations, and assume client won't provide any
+// identity certificates(single side TLS).
+class TlsChannelCredentialsOptions final : public TlsCredentialsOptions {
+ public:
+  // Sets the decision of whether to do a crypto check on the server certs.
+  // The default is true.
+  void set_verify_server_certs(bool verify_server_certs);
```

Comment:
Leaving a TODO for myself to go back through old notes - I believe we discussed this a while back.
I'll either remove it or add a discussion of how dangerous it is.

Reply:
Pinging this so we don't lose it.

Reply:
Looking back through old notes, this is not something we said we would remove - this PR https://github.com/grpc/grpc/pull/36488 marks those that will be removed.

In terms of what we should do here - IMO we should structure it differently. I propose the following:
`do_not_verify_server_certificates_dangerous()`

(Of course with more comments as well). I think this makes it very clear that you must _explicitly_ say "I don't want to verify server certs" rather than it just being a boolean option.
Because realistically, almost every use case will have to call `set_verify_server_certificates(true)`, so it would be confusing to mark it dangerous. Instead we can make that the default, then have _only_ a dangerous marked API to not verify the certs.

WDYT?

Reply:
SGTM. Consider `insecure_skip_server_certificate_verification()` to match Golang? Up to you, as long as there's some warning in the name.

Reply:
I lean a little towards `_dangerous` to keep it consistent with the other dangerous fns in this API, WDYT?
I'd rather be internally consistent with other C++ APIs than externally consistent with Go?